src/Compilers/CSharp/Portable/Parser/Lexer_RawStringLiteral.cs

```diff
@@ -0,0 +1,347 @@
+        private int GetNewLineWidth(char currentChar)
+        {
+            Debug.Assert(SyntaxFacts.IsNewLine(currentChar));
+            return currentChar == '\r' && TextWindow.PeekChar(1) == '\n' ? 2 : 1;
+        }
+
+        private void ScanMultiLineRawStringLiteral(ref TokenInfo info, int startingQuoteCount)
+        {
+            info.Kind = SyntaxKind.MultiLineRawStringLiteralToken;
+
+            // The indentation-whitespace computed from the very last line of the raw string literal
+            var indentationWhitespace = PooledStringBuilder.GetInstance();
+
+            // The leading whitespace of whatever line we are currently on.
+            var currentLineWhitespace = PooledStringBuilder.GetInstance();
+            try
+            {
+                // Do the first pass, finding the end of the raw string, and determining the 'indentation whitespace'
+                // that must be complimentary with all content lines of the raw string literal.
+                var afterStartDelimeter = TextWindow.Position;
+                Debug.Assert(SyntaxFacts.IsNewLine(TextWindow.PeekChar()));
+
+                var contentLineCount = 0;
+                while (ScanMultiLineRawStringLiteralLine(startingQuoteCount, indentationWhitespace.Builder))
+                    contentLineCount++;
+
+                // If the initial scan failed then just bail out without a constant value.
+                if (this.HasErrors)
+                    return;
+
+                // The trivial raw string literal is not legal in the language.
+                if (contentLineCount == 0)
+                {
+                    this.AddError(
+                        position: TextWindow.Position - startingQuoteCount,
+                        width: startingQuoteCount,
+                        ErrorCode.ERR_Multi_line_raw_string_literals_must_contain_at_least_one_line_of_content);
+                    return;
+                }
+
+                // Now, do the second pass, building up the literal value.  This may produce an error as well if the
+                // indentation whitespace of the lines isn't complimentary.
+
+                // Reset us to right after the starting delimiter.  Note: if we fail to generate a constant value we'll
+                // ensure that we reset back to the original end we scanned to above.
+                var tokenEnd = TextWindow.Position;
+                TextWindow.Reset(afterStartDelimeter);
+                Debug.Assert(SyntaxFacts.IsNewLine(TextWindow.PeekChar()));
+
+                for (var currentLine = 0; currentLine < contentLineCount; currentLine++)
+                {
+                    AddMultiLineRawStringLiteralLineContents(
+                        indentationWhitespace.Builder,
+                        currentLineWhitespace.Builder,
+                        firstContentLine: currentLine == 0);
+
+                    // If processing the line produced errors, then bail out from continued processing.
```

Comment:
Should we use `tokenEnd` here too? (error scenario)

Reply:
not sure what you mean.  can you clarify?

Reply:
My bad. I misread the flow. This is fine